sdk/core/Azure.Core.Experimental/src/DynamicData/DynamicData.Converters.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Azure.Core.Dynamic
+{
+    public partial class DynamicData
+    {
+        // From: https://github.com/Azure/autorest.csharp/blob/d835b0b7bffae08c1037ccc5824e928eaac55b96/src/assets/Generator.Shared/TypeFormatters.cs#L14
+        private const string RoundtripZFormat = "yyyy-MM-ddTHH:mm:ss.fffffffZ";
+
+        private class Rfc3339DateTimeConverter : JsonConverter<DateTime>
+        {
+            public override DateTime Read(
+                ref Utf8JsonReader reader,
+                Type typeToConvert,
+                JsonSerializerOptions options)
+            {
+                string? value = reader.GetString() ??
+                    throw new JsonException($"Failed to read 'string' value at JSON position {reader.Position}.");
+
+                // From: https://github.com/Azure/autorest.csharp/blob/d835b0b7bffae08c1037ccc5824e928eaac55b96/src/assets/Generator.Shared/TypeFormatters.cs#L130
+                DateTime dateTimeValue = DateTime.Parse(value, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal);
+                return dateTimeValue.ToUniversalTime();
+            }
+
+            public override void Write(
+                Utf8JsonWriter writer,
+                DateTime dateTimeValue,
+                JsonSerializerOptions options)
+            {
+                // From: https://github.com/Azure/autorest.csharp/blob/d835b0b7bffae08c1037ccc5824e928eaac55b96/src/assets/Generator.Shared/TypeFormatters.cs#LL19C84-L23C11
+                string value = dateTimeValue.Kind switch
+                {
+                    DateTimeKind.Utc => ((DateTimeOffset)dateTimeValue).ToUniversalTime().ToString(RoundtripZFormat, CultureInfo.InvariantCulture),
+                    _ => throw new NotSupportedException($"DateTime {dateTimeValue} has a Kind of {dateTimeValue.Kind}. Azure SDK requires it to be UTC. You can call DateTime.SpecifyKind to change Kind property value to DateTimeKind.Utc."),
```

Comment:
Can't we just convert from local time if `DateTimeKind.Local`? I can see throwing if it's not specified i.e., `DateTimeKind.Unspecified`, which isn't possible with `DateTimeOffset`.

Reply:
Might be good to at least ask if there was a good reason, but I don't think it should hold up this PR since you're just mimicking existing behavior. Thanks for the explanation!

Reply:
My guess for the reasoning is that the time zone of the `DateTime` is ambiguous if it's `DateTimeKind.Local` since it is dependent on the time zone of the system on which it was created.

Reply:
Presumably it's only on input, though. I'm thinking about client operations where the date-time is the local client time which we'd then adjust to UTC and serialize/submit. I wouldn't expect any non-UTC date to be coming from Azure and hopefully not coming into another service before calling to Azure. But I can see how this removes all ambiguity.

Reply:
This code might execute on the service (in Azure). I don't think we should be accepting local DateTimes. It's so error prone.